File: sdks/rust/src/application/ports.rs
//! `BackendAdapter` trait and the vendor-neutral resolution/target-
//! registration records it exchanges with
//! [`crate::application::runtime::FireweaveRuntime`].
//!
//! Adapters translate a backend (fw-server, an in-process dev map, in-memory
//! fixtures) into [`FlagResolution`] records. No vendor types cross this
//! boundary — the runtime and public API only ever see Fireweave-owned
//! shapes.
//!
//! All `BackendAdapter` methods take `&self` (not `&mut self`): a concrete
//! adapter is responsible for its OWN internal thread-safety (interior
//! mutability via `Mutex`/`RwLock`/atomics), exactly like go's
//! `BackendAdapter` interface and python's per-adapter locks — this lets
//! `FireweaveRuntime` hold `Box<dyn BackendAdapter>` directly and stay
//! `Send + Sync` without wrapping the whole adapter in an extra lock of its
//! own (the runtime's own lock — `application::runtime::RuntimeInner` —
//! guards only the lifecycle state machine and the context layers, mirroring
//! python's single `threading.RLock`).

use crate::domain::context::EvaluationContext;
use crate::domain::errors::{ErrorKind, FireweaveError};
use crate::domain::target::TargetKind;
use crate::domain::types::JsonValue;

/// `evaluate()`'s reserved fifth argument
/// (`conformance/surface/control-points.surface.json`:
/// `evaluate(key, type, default, context?, options?)`).
///
/// `include_payload` (`contracts/evaluation/eval-payload-attached.json`):
/// when true and the resolved flag carries a payload
/// ([`FlagResolution::payload`]), it is attached to
/// `flag_metadata["fireweave.payload"]` as a deterministic (sorted-key)
/// JSON string.
#[derive(Debug, Clone, Copy, Default)]
pub struct EvaluateOptions {
    pub include_payload: bool,
}

/// Options for `POST /v1/targets/register`
/// (`spec/remote-register-target.schema.json`).
///
/// Omitted fields are left off the wire rather than sent as null/default —
/// the server defaults `kind` to `user` when absent.
#[derive(Debug, Clone, Default)]
pub struct RegisterTargetOptions {
    pub kind: Option<TargetKind>,
    pub properties: Option<serde_json::Map<String, JsonValue>>,
    pub environment: Option<String>,
}

/// Outcome of target registration.
///
/// `ok: false` means the target was NOT registered — rules that depend on
/// its properties will not match until a later attempt succeeds. Callers in
/// a login path normally ignore this; a careful caller logs it — a silently
/// unregistered target is exactly how targeting rules end up matching
/// nobody.
#[derive(Debug, Clone)]
pub struct RegisterTargetResult {
    pub ok: bool,
    pub error: Option<FireweaveError>,
}

impl RegisterTargetResult {
    pub fn success() -> Self {
        RegisterTargetResult {
            ok: true,
            error: None,
        }
    }

    pub fn failure(error: FireweaveError) -> Self {
        RegisterTargetResult {
            ok: false,
            error: Some(error),
        }
    }
}

/// Vendor-neutral outcome of resolving one flag.
///
/// `matched: false` is the ONE typed channel an adapter uses to signal "no
/// decision for this key/context" back to the runtime — it is what
/// [`crate::application::runtime::FireweaveRuntime::evaluate`] reads to
/// produce reason `DEFAULT` (`spec/modes.md` "Behaviour per mode": local
/// mode's unknown-key row). Contrast a genuinely-unknown key at a real
/// backend (remote's "key unknown to the backend" row), which resolves to
/// reason `ERROR`/`FlagNotFound` by *returning `Err`* from `resolve`
/// instead of `Ok(FlagResolution { matched: false, .. })` — see
/// `infrastructure::adapters::remote::FireweaveRemoteAdapter::resolve` and
/// `infrastructure::adapters::local::FireweaveLocalAdapter::resolve`.
///
/// `fireweave_reason` lets an adapter force a canonical reason on a
/// *matched* resolution (e.g. the local dev adapter's `STATIC`).
#[derive(Debug, Clone)]
pub struct FlagResolution {
    pub value: JsonValue,
    pub variant: Option<String>,
    pub enabled: bool,
    pub matched: bool,
    pub version: Option<i64>,
    pub vendor_flag_id: Option<i64>,
    pub reason_code: Option<String>,
    pub condition_index: Option<i64>,
    pub payload: Option<JsonValue>,
    pub fireweave_reason: Option<String>,
    pub from_cache: bool,
    pub extra_metadata: serde_json::Map<String, JsonValue>,
}

impl Default for FlagResolution {
    fn default() -> Self {
        FlagResolution {
            value: JsonValue::Null,
            variant: None,
            enabled: true,
            matched: true,
            version: None,
            vendor_flag_id: None,
            reason_code: None,
            condition_index: None,
            payload: None,
            fireweave_reason: None,
            from_cache: false,
            extra_metadata: serde_json::Map::new(),
        }
    }
}

impl FlagResolution {
    /// A miss: no decision for this key/context
    /// (`spec/modes.md` "Behaviour per mode": local's unknown-key row).
    pub fn miss() -> Self {
        FlagResolution {
            matched: false,
            ..Default::default()
        }
    }
}

/// Protocol every Fireweave backend adapter implements.
///
/// `resolve` returns `Err(FireweaveError)` for a genuine backend failure
/// (`FlagNotFound`, `Network`, `Timeout`, ...); the runtime converts those
/// into default-valued decisions — evaluation APIs never propagate them to
/// the caller.
///
/// `register_target` carries a default implementation degrading with
/// `UnsupportedCapability` (mirroring go's `BackendAdapter.RegisterTarget`,
/// which every adapter implements — including the fixture-only in-memory
/// one, which degrades) rather than python's duck-typed optional method.
pub trait BackendAdapter: Send + Sync {
    /// Bring the backend up; return `Err` on fatal config.
    fn initialize(&self) -> Result<(), FireweaveError>;

    /// Resolve one flag against a validated, merged context.
    fn resolve(
        &self,
        flag_key: &str,
        context: &EvaluationContext,
    ) -> Result<FlagResolution, FireweaveError>;

    /// Deterministically release resources within `timeout_ms`.
    /// Idempotent; must never panic.
    fn shutdown(&self, timeout_ms: u64);

    /// Register a target. Adapters without the capability degrade with
    /// `UnsupportedCapability` (this default).
    fn register_target(
        &self,
        targeting_key: &str,
        options: Option<&RegisterTargetOptions>,
    ) -> RegisterTargetResult {
        let _ = (targeting_key, options);
        RegisterTargetResult::failure(FireweaveError::new(ErrorKind::UnsupportedCapability))
    }
}

File: sdks/dart/lib/src/application/ports.dart
/// Backend adapter boundary (mirrors `sdks/web`'s `WebBackendAdapter` and
/// swift's `ControlPointsBackendAdapter` — ADR-0009's seam).
///
/// One shape difference from node/go/java/rust's per-call `resolve(key)`
/// port is the whole reason this SDK can offer a synchronous `evaluate()`:
/// this port has `prefetch(context)`, which returns EVERY decision for a
/// context in one round trip. Evaluation then becomes a synchronous map
/// lookup in `FireweaveRuntime` — never an adapter call.
///
/// The reason is not performance: a Flutter `build()` cannot await, so
/// nothing on the read surface may need to suspend. Both
/// `FireweaveLocalAdapter` and `FireweaveRemoteAdapter` implement this SAME
/// port (`spec/modes.md`: "Both modes expose the identical nine methods with
/// identical signatures"), so `FireweaveRuntime`/`FireweaveClient` stay
/// mode-blind.
library;

import '../domain/context.dart';
import '../domain/decision.dart';
import '../domain/errors.dart';
import '../domain/target.dart';
import '../domain/types.dart';

/// Every decision the backend returned for one context, keyed by control
/// point.
typedef PrefetchResult = Map<String, AdapterResolution>;

/// Sink for the `[fireweave:local]` `registerTarget` trace line. Injectable
/// so tests assert the call without capturing stdout, and so a host that
/// owns its logging can route it (`spec/modes.md` "The ... log sink MUST be
/// injectable").
typedef LogSink = void Function(String message);

class PrefetchOptions {
  const PrefetchOptions({this.flagKeys});

  /// Restrict the batch to these keys; omit to let the backend return all
  /// it knows.
  final List<String>? flagKeys;
}

/// Vendor-neutral outcome of resolving one flag, as returned in a
/// [PrefetchResult] batch.
///
/// **[found] has a dual meaning that callers must not conflate**:
///
/// 1. **Present in the map with `found: false`** — the definition EXISTS but
///    its targeting conditions did not select this caller
///    (`InMemoryAdapter`'s matchAttribute/matchGroups/matchPerson/
///    matchTargetingKey). `FireweaveRuntime.evaluate` reads this as
///    `DecisionReason.defaultReason` UNCONDITIONALLY — "no decision for this
///    key/context" is a claim about the flag's own targeting, not about the
///    adapter's miss policy.
/// 2. **ABSENT from the map entirely** — governed by
///    [ControlPointsBackendAdapter.missReason]: `FireweaveLocalAdapter`
///    reports `defaultReason` here too (`spec/modes.md` "Behaviour per
///    mode": local mode's unknown-key row), while `FireweaveRemoteAdapter`/
///    `InMemoryAdapter` leave `missReason` `null` and an absent key resolves
///    to `ERROR`/`FlagNotFound` instead.
///
/// [vendorFlagId]/[reasonCode] are a PRE-GATED pair
/// (`spec/decision.schema.json` `standardMetadataKeys`, ruling 11): the
/// runtime emits `fireweave.vendorFlagId`/`fireweave.reasonCode` together,
/// or neither — never one alone. The gate is applied where the raw
/// "condition index" signal exists as adapter input (`InMemoryAdapter`),
/// before this resolution is constructed.
class AdapterResolution {
  const AdapterResolution({
    this.found = true,
    this.enabled,
    this.value,
    this.variant,
    this.flagType,
    this.reason,
    this.reasonCode,
    this.version,
    this.vendorFlagId,
    this.payload,
    this.fromCache = false,
  });

  final bool found;
  final bool? enabled;
  final JsonValue value;
  final String? variant;
  final FlagType? flagType;
  final DecisionReason? reason;
  final String? reasonCode;
  final int? version;
  final int? vendorFlagId;
  final JsonValue payload;
  final bool fromCache;

  /// Plain-JSON snapshot; used to detect changed keys across prefetches.
  Map<String, Object?> toJson() => <String, Object?>{
    'found': found,
    'enabled': enabled,
    'value': value,
    'variant': variant,
    'flagType': flagType?.wireName,
    'reason': reason?.wireName,
    'reasonCode': reasonCode,
    'version': version,
    'vendorFlagId': vendorFlagId,
    'payload': payload,
    'fromCache': fromCache,
  };
}

class RegisterTargetOptions {
  const RegisterTargetOptions({this.kind, this.properties, this.environment});

  final TargetKind? kind;
  final Map<String, Object?>? properties;
  final String? environment;
}

/// Outcome of target registration.
///
/// `ok: false` means the target was NOT registered — rules that depend on
/// its properties will not match until a later attempt succeeds. Callers in
/// a sign-in path normally ignore this; a careful caller logs it — a
/// silently unregistered target is exactly how targeting rules end up
/// matching nobody.
class RegisterTargetResult {
  const RegisterTargetResult._({required this.ok, this.error});

  const RegisterTargetResult.success() : this._(ok: true);

  const RegisterTargetResult.failure(FireweaveError error)
    : this._(ok: false, error: error);

  final bool ok;
  final FireweaveError? error;
}

/// `evaluate()`'s reserved fifth argument
/// (`conformance/surface/control-points.surface.json`:
/// `evaluate(key, type, default, context?, options?)`).
///
/// [includePayload] is FUNCTIONAL here, as in swift (and unlike web's inert
/// options): payload attachment is genuine v1 surface
/// (`Decision.flagMetadata['fireweave.payload']`), this SDK runs the shared
/// 65 fixtures for real (`eval-payload-attached`), and
/// [AdapterResolution.payload] already carries the raw payload from prefetch
/// to the read.
class EvaluateOptions {
  const EvaluateOptions({this.includePayload = false});

  final bool includePayload;
}

/// One HTTP response as seen by the remote adapter.
class TransportResponse {
  const TransportResponse({required this.statusCode, required this.body});

  final int statusCode;
  final String body;
}

/// Injectable HTTP transport — the ONE place this SDK touches the network.
///
/// A platform default is chosen by conditional import on every Dart target:
/// `dart:io`'s `HttpClient` on the VM and on Flutter for Android, iOS,
/// macOS, Windows, and Linux; the browser's `fetch` (via `dart:js_interop`)
/// on Flutter web and under `dart compile js`/`wasm`. Tests inject a fake;
/// an app with its own HTTP stack can inject that instead.
///
/// Implementations MUST throw a [FireweaveError] of kind `Timeout` or
/// `Network` for transport-level failures; the adapter maps HTTP status
/// codes itself.
abstract interface class HttpTransport {
  Future<TransportResponse> post(
    Uri url, {
    required Map<String, String> headers,
    required String body,
    required Duration timeout,
  });
}

/// Protocol every Fireweave backend adapter implements.
abstract interface class ControlPointsBackendAdapter {
  /// Miss-reason override for a control point ABSENT from the prefetch
  /// result (`spec/modes.md` "Behaviour per mode": local mode's unknown-key
  /// row is `default`/reason `DEFAULT`, not an error — unlike remote's
  /// `default`/`ERROR`/`FlagNotFound`). `FireweaveLocalAdapter` returns
  /// `DecisionReason.defaultReason` here; `FireweaveRemoteAdapter` and
  /// `InMemoryAdapter` return `null` and keep the FlagNotFound/ERROR path.
  DecisionReason? get missReason;

  /// Bring the backend to a usable state. Throws [FireweaveError] on fatal
  /// config.
  Future<void> initialize();

  /// Fetch every decision for a context in one round trip. Throws
  /// [FireweaveError] on transport faults.
  Future<PrefetchResult> prefetch(
    EvaluationContext context, {
    PrefetchOptions? options,
  });

  /// Register a target. Resolves rather than throws — this runs in sign-in
  /// paths, where a targeting concern must not break authentication.
  Future<RegisterTargetResult> registerTarget(
    String targetingKey, {
    RegisterTargetOptions? options,
  });

  /// Deterministically release resources. Idempotent; must never throw.
  Future<void> shutdown();
}
